README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ggdirectlabel

<!-- badges: start -->

[![R-CMD-check](https://github.com/MattCowgill/ggdirectlabel/workflows/R-CMD-check/badge.svg)](https://github.com/MattCowgill/ggdirectlabel/actions)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R-CMD-check](https://github.com/MattCowgill/ggdirectlabel/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/MattCowgill/ggdirectlabel/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The goal of ggdirectlabel is to make it easier to directly label ggplot2
charts rather than using legends.

## Installation

You can install the development version of ggdirectlabel from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("MattCowgill/ggdirectlabel")
```

``` r
library(ggdirectlabel)
library(ggplot2)
library(magrittr)
```

## Using `geom_richlegend()`

Here’s a standard ggplot2 scatterplot:

``` r
base_scatter <- mtcars |> 
  ggplot(aes(x = wt, y = mpg, col = factor(cyl))) +
  geom_point()

base_scatter
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="672" />

This is fine! But sometimes you might like the legend levels (4, 6, and
8 in this example) to be coloured according to the levels in the data.
That’s where `geom_richlegend()` comes in:

``` r
base_scatter +
  geom_richlegend(aes(label = cyl)) +
  theme(legend.position = "none")
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="672" />

You can move the ‘rich legend’ around:

``` r
base_scatter +
  geom_richlegend(aes(label = cyl),
                  legend.position = "bottomleft",
                  vjust = 0,
                  hjust = 0) +
  theme(legend.position = "none")
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="672" />

`geom_richlegend()` respects facets - it’ll place a little legend
annotation for each level of the data that appears in that panel:

``` r
base_scatter +
  geom_richlegend(aes(label = paste0(cyl, " cylinders"))) +
  facet_wrap(~cyl)
```

<img src="man/figures/README-unnamed-chunk-6-1.png" width="672" />

## Using `geom_linepoint()`

Without ggirectlabel, we might do something like:

``` r

ggplot2::economics_long %>%
  ggplot(aes(x = date, y = value, col = variable)) +
  geom_line() +
  geom_point(data = ~dplyr::filter(., date == max(date)),
             fill = "white",
             shape = 21,
             size = 2.5,
             stroke = 1.25)
```

<img src="man/figures/README-no-directlabel-1.png" width="672" />

This is fine! But this is a more straightforward way to achieve the same
thing:

``` r
ggplot2::economics_long %>%
  ggplot(aes(x = date, y = value, col = variable)) +
  geom_linepoint()
```

<img src="man/figures/README-example-1.png" width="672" />

## Using `scale_x_date_rightalign()`

In time series line charts, it’s often important to make clear the date
of your most recent observation. The `scale_x_date_rightalign()`
function aligns the breaks of your x-axis so that the most recent
observation is included in the breaks.

``` r
ggplot2::economics_long %>%
  ggplot(aes(x = date, y = value, col = variable)) +
  geom_linepoint() +
  scale_x_date_rightalign()
```

<img src="man/figures/README-scale_x_date_rightalign-1.png" width="672" />

## Using `geom_finallabel()`

In time series line charts, you may wish to label the final point in the
series. The `geom_finallabel()` function makes that easy.

``` r
ggplot2::economics_long %>%
  ggplot(aes(x = date, y = value, col = variable)) +
  geom_linepoint() +
  geom_finallabel(aes(label = round(value, 0))) +
  scale_x_date_rightalign(expand = expansion(c(0, 0.15))) +
  theme(legend.position = "none")
```

<img src="man/figures/README-geom_finallabel-1.png" width="672" />


NEWS.md
# ggdirectlabel 0.1.0.9xx
* Bug fixes

# ggdirectlabel 0.1.0
* geom_richlegend() added

# ggdirectlabel 0.0.1
* initial working version, with scale_x_date_*align(), breaks_right(), 
geom_linepoint() and geom_finallabel() 


tests/testthat.R
library(testthat)
library(ggdirectlabel)

test_check("ggdirectlabel")


tests/testthat/test-geom.R
test_that("geom_linepoint() produces expected output", {
  library(ggplot2)

  plot <-  ggplot2::economics_long %>%
      ggplot(aes(x = date, y = value)) +
      geom_linepoint(aes(col = variable)) +
      facet_wrap(~variable)

  expect_s3_class(plot, "gg")
  vdiffr::expect_doppelganger("Faceted plot", plot)

  vdiffr::expect_doppelganger("Numeric axes",
                              iris %>%
    ggplot(aes(x = Sepal.Length, y = Sepal.Width, col = Species)) +
    geom_linepoint())

  vdiffr::expect_doppelganger("No colour",
  ggplot2::economics_long %>%
    ggplot(aes(x = date, y = value)) +
    geom_linepoint() +
    facet_wrap(~variable) )

})


tests/testthat/test-geom_linepoint.R
library(ggplot2)

test_that("geom_linepoint() produces expected output", {

  plot <-  ggplot(ggplot2::economics_long, aes(x = date, y = value)) +
      geom_linepoint(aes(col = variable)) +
      facet_wrap(~variable, scales = "free_y")

  expect_s3_class(plot, "gg")
  vdiffr::expect_doppelganger("Faceted plot", plot)

  vdiffr::expect_doppelganger("Numeric axes",
    ggplot(iris,
           aes(x = Sepal.Length, y = Sepal.Width, col = Species)) +
    geom_linepoint())

  vdiffr::expect_doppelganger("No colour",
    ggplot(ggplot2::economics_long,
           aes(x = date, y = value)) +
    geom_linepoint() +
    facet_wrap(~variable) )

})


test_that("non-ordered data displays correctly with geom_linepoint()", {
  plot <- economics[order(-economics$unemploy), , drop = FALSE] |>
    ggplot(aes(x = date, y = unemploy)) +
    geom_linepoint()

  vdiffr::expect_doppelganger("Unordered data", plot)
})


R/geom.R
#' Line with a dot on the final observation
#' @description `geom_linepoint()` draws a `ggplot2::geom_line()` and adds a
#' `ggplot2::geom_point()` at the observation with the maximum x value for each
#' group.
#' @inheritParams ggplot2::geom_line
#' @section Aesthetics:
#' \code{geom_LinePoint()} understands the following aesthetics (required aesthetics are in bold):
#'   \itemize{
#'     \item \strong{\code{x}}
#'     \item \strong{\code{y}}
#'     \item \code{alpha}
#'     \item \code{colour}
#'     \item \code{group}
#'     \item \code{linetype}
#'     \item \code{pointfill}
#'     \item \code{pointshape}
#'     \item \code{pointsize}
#'     \item \code{pointstroke}
#'     \item \code{pointshape}
#'     \item \code{size}
#'     \item \code{weight}
#'   }
#' The aesthetics that begin with 'point' (eg. `pointfill`) are passed to
#' `geom_point()` - for example `pointfill` is passed to the `fill` aesthetic
#' of `geom_point()`.
#'
#' The `x`, `y`, `alpha`, `colour`, and `group` aesthetics are passed to both
#' `geom_line()` and `geom_point()`.
#'
#' The `linetype`, `size`, and `weight` aesthetics are passed to `geom_line()`.
#'
#' Learn more about setting these aesthetics in \code{vignette("ggplot2-specs")}.
#' @seealso `ggplot2::geom_line`, `ggplot2::geom_point`
#' @export
#' @examples
#' library(ggplot2)
#'
#' ggplot2::economics_long %>%
#'   ggplot(aes(x = date, y = value)) +
#'   geom_linepoint(aes(col = variable)) +
#'   facet_wrap(~variable)
geom_linepoint <-
  function(mapping = NULL,
           data = NULL,
           stat = "identity",
           position = "identity",
           na.rm = FALSE,
           show.legend = NA,
           inherit.aes = TRUE,
           ...) {
    ggplot2::layer(
      data = data,
      mapping = mapping,
      stat = stat,
      geom = GeomLinePoint,
      position = position,
      show.legend = show.legend,
      inherit.aes = inherit.aes,
      params = list(
        na.rm = na.rm,
        ...
      )
    )
  }

GeomLinePoint <- ggplot2::ggproto(
  "GeomLinePoint",
  ggplot2::Geom,
  extra_params = c("na.rm"),
  setup_data = function(data, params) {
    ggplot2::GeomPoint$setup_data(data, params)
  },
  draw_group = function(data,
                        panel_params,
                        coord,
                        lineend = "butt",
                        linejoin = "round",
                        linemitre = 10,
                        size = 5,
                        flipped_aes = FALSE) {
    point <- transform(data)
    point <- point[point$x == max(point$x), ]
    point$size <- point$pointsize
    point$fill <- point$pointfill
    point$shape <- point$pointshape
    point$stroke <- point$pointstroke
    point$shape <- point$pointshape

    path <- transform(data, alpha = NA)

    grid::gList(
      ggplot2::GeomLine$draw_panel(
        path,
        panel_params,
        coord,
        lineend = lineend,
        linejoin = linejoin,
        linemitre = linemitre
      ),
      ggplot2::GeomPoint$draw_panel(point, panel_params, coord)
    )
  },
  draw_key = ggplot2::draw_key_smooth,
  required_aes = c("x", "y"),
  non_missing_aes = c(
    "size", "shape", "colour", "pointsize",
    "pointstroke", "pointfill", "pointshape"
  ),
  default_aes = ggplot2::aes(
    pointsize = 2.5, pointfill = "white", pointshape = 21,
    shape = 19, colour = "black", size = 1,
    alpha = 1, pointstroke = 1.5, linetype = 1, weight = 1
  )
)


R/geom_linepoint.R
#' Line with a dot on the final observation
#' @description `geom_linepoint()` draws a `ggplot2::geom_line()` and adds a
#' `ggplot2::geom_point()` at the observation with the maximum x value for each
#' group.
#' @inheritParams ggplot2::geom_line
#' @section Aesthetics:
#' \code{geom_linepoint()} understands the following aesthetics (required aesthetics are in bold):
#'   \itemize{
#'     \item \strong{\code{x}}
#'     \item \strong{\code{y}}
#'     \item \code{alpha}
#'     \item \code{colour}
#'     \item \code{group}
#'     \item \code{linetype}
#'     \item \code{pointfill}
#'     \item \code{pointshape}
#'     \item \code{pointsize}
#'     \item \code{pointstroke}
#'     \item \code{pointshape}
#'     \item \code{linewidth}
#'     \item \code{weight}
#'   }
#' The aesthetics that begin with 'point' (eg. `pointfill`) are passed to
#' `geom_point()` - for example `pointfill` is passed to the `fill` aesthetic
#' of `geom_point()`.
#'
#' The `x`, `y`, `alpha`, `colour`, and `group` aesthetics are passed to both
#' `geom_line()` and `geom_point()`.
#'
#' The `linetype`, `linewidth`, and `weight` aesthetics are passed to `geom_line()`.
#'
#' Learn more about setting these aesthetics in \code{vignette("ggplot2-specs")}.
#' @seealso `ggplot2::geom_line`, `ggplot2::geom_point`
#' @rdname geom_linepoint
#' @export
#' @examples
#' library(ggplot2)
#'
#'ggplot(ggplot2::economics_long, aes(x = date, y = value)) +
#'   geom_linepoint(aes(col = variable)) +
#'   facet_wrap(~variable)
geom_linepoint <-
  function(mapping = NULL,
           data = NULL,
           stat = "identity",
           position = "identity",
           na.rm = FALSE,
           show.legend = NA,
           inherit.aes = TRUE,
           ...) {
    ggplot2::layer(
      data = data,
      mapping = mapping,
      stat = stat,
      geom = GeomLinePoint,
      position = position,
      show.legend = show.legend,
      inherit.aes = inherit.aes,
      params = list(
        na.rm = na.rm,
        ...
      )
    )
  }

#' @rdname geom_linepoint
#' @export
GeomLinePoint <- ggplot2::ggproto(
  "GeomLinePoint",
  ggplot2::Geom,
  extra_params = c("na.rm"),
  setup_data = function(data, params) {
    ggplot2::GeomLine$setup_data(data, params)
  },
  draw_group = function(data,
                        panel_params,
                        coord,
                        lineend = "butt",
                        linejoin = "round",
                        linemitre = 10,
                        size = 5,
                        flipped_aes = FALSE) {
    point <- data
    point <- point[point$x == point$x[length(point$x)], ]
    point$size <- point$pointsize
    point$fill <- point$pointfill
    point$shape <- point$pointshape
    point$stroke <- point$pointstroke
    point$shape <- point$pointshape

    if (utils::packageVersion("ggplot2") < "3.4.0") {
      names(data)[names(data) == "linewidth"] <- "size"
    }

    path <- transform(data, alpha = NA)

    grid::gList(
      ggplot2::GeomLine$draw_panel(
        path,
        panel_params,
        coord,
        lineend = lineend,
        linejoin = linejoin,
        linemitre = linemitre
      ),
      ggplot2::GeomPoint$draw_panel(point, panel_params, coord)
    )
  },
  draw_key = ggplot2::draw_key_smooth,
  required_aes = c("x", "y"),
  non_missing_aes = c(
    "linewidth", "shape", "colour", "pointsize",
    "pointstroke", "pointfill", "pointshape"
  ),
  default_aes = ggplot2::aes(
    pointsize = 2.5,
    pointfill = "white",
    pointshape = 21,
    shape = 19,
    colour = "black",
    linewidth = 1,
    alpha = 1,
    pointstroke = 1.5,
    linetype = 1,
    weight = 1
  )
)


R/geom_finallabel.R
#' Line with a dot on the final observation
#' @description `geom_finallabel()` draws a `ggplot2::geom_text()`
#' at the observation with the maximum x value for each group.
#' @inheritParams ggplot2::geom_text
#' @param nudge_x_perc Amount to nudge label along the x-axis, expressed as a
#' percentage of the data range along the x-axis. The default is `1.5`, which
#' will nudge the text 1.5% of the data range from the final point. This is applied
#' in addition to `nudge_x`, which nudges the text a specific number of data units.
#' @section Aesthetics:
#'   \code{geom_text()} understands the following aesthetics (required aesthetics are in bold):
#'   \itemize{
#'     \item \strong{\code{x}}
#'     \item \strong{\code{y}}
#'     \item \strong{\code{label}}
#'     \item \code{alpha}
#'     \item \code{angle}
#'     \item \code{colour}
#'     \item \code{family}
#'     \item \code{fontface}
#'     \item \code{group}
#'     \item \code{hjust}
#'     \item \code{lineheight}
#'     \item \code{size}
#'     \item \code{vjust}
#'   }
#'
#' Learn more about setting these aesthetics in \code{vignette("ggplot2-specs")}.
#' @seealso `ggplot2::geom_text`
#' @export
#' @rdname geom_finallabel
#' @examples
#' library(ggplot2)
#'
#'ggplot(ggplot2::economics_long, aes(x = date, y = value)) +
#'   geom_linepoint(aes(col = variable)) +
#'   geom_finallabel(aes(label = value)) +
#'   facet_wrap(~variable)
geom_finallabel <-
  function(mapping = NULL,
           data = NULL,
           stat = "identity",
           position = "identity",
           nudge_x = 0,
           nudge_y = 0,
           nudge_x_perc = 1.5,
           na.rm = FALSE,
           show.legend = FALSE,
           inherit.aes = TRUE,
           ...) {
    if (!missing(nudge_x) || !missing(nudge_y)) {
      if (!missing(position)) {
        stop("You must specify either `position` or `nudge_x`/`nudge_y`.")
      }

      position <- position_nudge(nudge_x, nudge_y)
    }

    ggplot2::layer(
      data = data,
      mapping = mapping,
      stat = stat,
      geom = GeomFinalLabel,
      position = position,
      show.legend = show.legend,
      inherit.aes = inherit.aes,
      params = list(
        na.rm = na.rm,
        nudge_x_perc = nudge_x_perc,
        ...
      )
    )
  }

#' @export
#' @rdname geom_finallabel
GeomFinalLabel <- ggplot2::ggproto(
  "GeomFinalLabel",
  ggplot2::Geom,
  extra_params = c("na.rm", "nudge_x_perc"),
  setup_data = function(data, params) {
    ggplot2::GeomText$setup_data(data, params)
  },
  draw_group = function(data,
                        panel_params,
                        coord,
                        nudge_x_perc,
                        flipped_aes = FALSE) {
    x_range <- range(data$x)
    x_min <- x_range[1]
    x_max <- x_range[2]
    data <- data[data$x == x_max, ]
    data$x <- data$x + ((x_max - x_min) * (nudge_x_perc / 100))


    ggplot2::GeomText$draw_panel(
        data,
        panel_params,
        coord
      )
  },
  draw_key = ggplot2::draw_key_text,
  required_aes = c("x", "y", "label"),
  default_aes = ggplot2::aes(
    colour = ggplot2::GeomText$default_aes$colour,
    size = ggplot2::GeomText$default_aes$size,
    angle = ggplot2::GeomText$default_aes$angle,
    hjust = 0,
    vjust = ggplot2::GeomText$default_aes$vjust,
    alpha = ggplot2::GeomText$default_aes$alpha,
    family = ggplot2::GeomText$default_aes$family,
    fontface = ggplot2::GeomText$default_aes$fontface,
    lineheight = 0.9
  )
)


R/scale_x_date_rightalign.R
#' Add a ggplot2 date scale with breaks aligned to the end of your data
#'
#' When visualising time series, it's often important to be clear about
#' the date of the latest observation. `scale_x_date_rightalign()` aligns
#' your axis breaks to the most recent observation in your data.
#'
#' `scale_x_date_leftalign()` aligns your breaks to the initial date in
#' your data.
#'
#' `scale_x_date_bothalign()` ensures that your breaks include both the initial
#' and final date in your data, with evenly-spaced breaks in between.
#'
#' @param expand The amount of padding/expansion that should be added at the
#' left and right of the data. Use the `ggplot2::expansion()` function to
#' add padding. By default, 5% padding is added to both right and left.
#' @param num_breaks The (approximate) number of breaks to include on the
#' axis. `base::pretty()` is used to generate these breaks.
#' @param date_labels The format for the date labels on the axis. The default
#' means "day of month, then shortened month, then linebreak, then
#' full year". See `?strptime` for codes that can be used here. The default is
#' `NULL`, which means a sensible default will be inferred from your data.
#' @param ... Arguments passed to `ggplot2::scale_x_date()`.
#' @export
#' @rdname scale_date_align
#' @import ggplot2
#' @examples
#' library(ggplot2)
#' ggplot(ggplot2::economics, aes(date, unemploy)) +
#'     geom_line() +
#'     scale_x_date_rightalign()
#'
#' ggplot(ggplot2::economics, aes(date, unemploy)) +
#'     geom_line() +
#'     scale_x_date_leftalign()
#'
#' ggplot(ggplot2::economics, aes(date, unemploy)) +
#'     geom_line() +
#'     scale_x_date_bothalign()
scale_x_date_rightalign <- function(expand = expansion(mult = c(0.05, 0.1)),
                                    num_breaks = 5,
                                    date_labels = NULL,
                                    ...) {
  scale_x_date_align(expand = expand,
                     num_breaks = num_breaks,
                     date_labels = date_labels,
                     align = "right",
                     ...)
}

#' @rdname scale_date_align
#' @export
scale_x_date_leftalign <- function(expand = expansion(mult = c(0.05, 0.05)),
                                   num_breaks = 5,
                                   date_labels = NULL,
                                   ...) {
  scale_x_date_align(expand = expand,
                     num_breaks = num_breaks,
                     date_labels = date_labels,
                     align = "left",
                     ...)
}

#' @rdname scale_date_align
#' @export
scale_x_date_bothalign <- function(expand = expansion(mult = c(0.05, 0.05)),
                                   num_breaks = 5,
                                   date_labels = NULL,
                                   ...) {
  scale_x_date_align(expand = expand,
                     num_breaks = num_breaks,
                     date_labels = date_labels,
                     align = "both",
                     ...)
}

scale_x_date_align <- function(expand = expansion(mult = c(0.05, 0.05)),
                               num_breaks = 5,
                               date_labels = NULL,
                               align = c("both", "right", "left"),
                               ...) {

  align <- match.arg(align)

  date_labeller <- if (is.null(date_labels)) {
    function(x, fmt = labels_date_auto(x)) format(x, fmt)
  } else {
    function(x, fmt = date_labels) format(x, fmt)
  }

  scale_x_date(
    expand = expand,
    breaks = function(limits,
                      exp = expand,
                      break_num = num_breaks,
                      date_align = align) {
      exp_mult <- exp[c(1, 3)]
      exp_add <- exp[c(2, 4)]

      range_perc_data <- 1 + exp_mult[1] + exp_mult[2]
      data_range <- as.numeric((1 / range_perc_data) *
                                 (limits[2] - limits[1] - exp_add[1] - exp_add[2]))

      data_max <-
        limits[2] - (data_range * (exp_mult[2]))  - exp_add[2]

      data_min <- data_max - data_range

      if (date_align == "right") {
        breaks <- breaks_right(limits = c(data_min, data_max),
                               n_breaks = break_num)
      } else if (date_align == "left") {
        breaks <- breaks_left(limits = c(data_min, data_max),
                               n_breaks = break_num)
      } else {
        breaks <- seq.Date(from = data_min,
                           to = data_max,
                           length.out = break_num)
      }



      breaks
    },
    labels = date_labeller,
    ...
  )
}


#' Generate date breaks that align with the maximum data value
#'
#' This function generates a vector of of 'pretty'
#' breaks (using `scales::breaks_pretty()`) that ends with
#' the upper limit provided and excludes any values that lie
#' outside the limits.
#'
#' @param limits Length-two numeric or date vector
#' @param n_breaks Number of breaks; passed to `scales::breaks_pretty()`
#' @param ... Passed to `scales::breaks_pretty()`
#' @export
#' @seealso scale_x_date_rightalign()
#' @return Vector of breaks, of the same class as `limits`
#' @author Originally written by Matt Cowgill for the `djprtheme` package,
#' in which the Department of Jobs, Precincts and Regions (Victoria) is
#' the copyright holder.
#' @examples
#'# Can be used with numeric vectors
#'breaks_right(c(10, 30))
#'
#'# Or date vectors
#'econ_dates <- range(ggplot2::economics$date)
#'breaks_right(econ_dates)
#'
#'# Can be supplied directly to the `breaks` argument of
#'# `ggplot2::scale_*_continuous()`. Note that the breaks will
#'# be aligned to the right-limit of the plot
#'# (including expansion/padding area).
#'
#'library(ggplot2)
#'
#'ggplot(ggplot2::economics,
#'       aes(x = date, y = unemploy)) +
#'  geom_line() +
#'  scale_x_date(breaks = breaks_right)
#'
#' # To right-align to the data, not including padding, try:
#' ggplot(ggplot2::economics,
#'     aes(x = date, y = unemploy)) +
#'     geom_line() +
#'     scale_x_date(breaks = breaks_right(limits = range(
#'     ggplot2::economics$date)))
#'
#' # Or use `scale_x_date_rightalign()`:
#' ggplot(ggplot2::economics,
#'     aes(x = date, y = unemploy)) +
#'     geom_line() +
#'     scale_x_date_rightalign()
#'
breaks_right <- function(limits,
                         n_breaks = 5,
                         ...) {
  breaks_align(limits = limits,
               n_breaks = n_breaks,
               date_align = "right",
               ...)
}

breaks_left <- function(limits,
                        n_breaks = 5,
                        ...) {
  breaks_align(limits = limits,
               n_breaks = n_breaks,
               date_align = "left",
               ...)
}

breaks_align <- function(limits,
                         n_breaks,
                         date_align = c("right", "left"),
                         ...) {
  date_align <- match.arg(date_align)
  min_date <- limits[1]
  max_date <- limits[2]
  pre_br <- scales::breaks_pretty(n = n_breaks,
                                  min.n = n_breaks - 1,
                                  ...)(c(min_date, max_date))


  if (date_align == "right") {
    date_adj <- as.numeric(pre_br[length(pre_br)] - max_date)
  } else {
    date_adj <- as.numeric(min(pre_br[pre_br >= min_date]) - min_date)
  }

  adj_br <- pre_br - date_adj
  names(adj_br) <- NULL
  out_br <- adj_br[adj_br >= min_date & adj_br <= max_date]
  out_br
}

labels_date_auto <- function(dates) {
  date_range <- as.numeric(max(dates) - min(dates))
  days_per_obs <- date_range / length(unique(dates))

  if (days_per_obs < 28) {
    fmt <- "%e %b\n%Y"
  } else if (days_per_obs < 365) {
    fmt <- "%b\n%Y"
  } else {
    fmt <- "%Y"
  }
  fmt
}


R/geom_richlegend.R
#' Create a 'rich legend' for your ggplot2 plot.
#' @description `geom_richlegend()` draws coloured text in lieu of a legend. It
#' uses the wonderful `{ggtext}` to create coloured text annotation(s) at a
#' location of your choice on your plot. Use this instead of a standard legend.
#' @param legend.position Either:
#'
#'  - a two-element numeric vector such as `c(0.2, 0.9)`. The first element
#'  denoted the x-position of the data and the second element denotes the
#'  y-position. Each element must be between 0 and 1 (inclusive).
#'
#'  - one of the following strings: "left", "right", "bottom", "top",
#'  "topleft", "topright", "bottomleft", "bottomright".
#'
#'  Default is "topright", which is equivalent to `c(0.975, 0.975)`.
#'
#' @inheritParams ggtext::geom_richtext
#' @examples
#' library(ggplot2)
#'
#' base_plot <- ggplot(mtcars, aes(x = wt, y = mpg, col = factor(cyl))) +
#'   geom_point() +
#'   theme(legend.position = "none")
#'
#' # By default, the rich legend is placed at the "topright"
#' base_plot +
#'   geom_richlegend(aes(label = cyl))
#'
#' # You can change the position of the rich legend:
#' base_plot +
#'   geom_richlegend(aes(label = cyl),
#'                   legend.position = "top")
#'
#' # Or you can change the position using a numeric vector:
#' base_plot +
#'   geom_richlegend(aes(label = cyl),
#'                   legend.position = c(0.1, 0.1))
#'
#' # The legend respects facets:
#' base_plot +
#'   geom_richlegend(aes(label = cyl)) +
#'   facet_wrap(~cyl)
#'
#'
#' @import ggplot2
#' @rdname geom_richlegend
#' @export
geom_richlegend <-
  function(mapping = NULL,
           data = NULL,
           legend.position = "topright",
           na.rm = FALSE,
           inherit.aes = TRUE,
           ...) {
    ggplot2::layer(
      data = data,
      mapping = mapping,
      stat = "identity",
      geom = GeomRichLegend,
      position = "identity",
      show.legend = FALSE,
      inherit.aes = inherit.aes,
      params = list(
        na.rm = na.rm,
        legend.position = legend.position,
        ...
      )
    )
  }

#' @rdname geom_richlegend
#' @export
#' @importFrom ggtext GeomRichText
#' @importFrom dplyr mutate
GeomRichLegend <- ggplot2::ggproto(
  "GeomRichLegend",
  ggplot2::Geom,
  extra_params = c("na.rm",
                   "legend.position"),
  setup_data = function(data, params) {

    out <- ggtext::GeomRichText$setup_data(data, params) |>
      dplyr::group_by(label, PANEL, colour) |>
      dplyr::summarise() |>
      dplyr::distinct() |>
      dplyr::ungroup() |>
      dplyr::mutate(legend.position = list(params$legend.position))

    out
  },
  draw_panel = function(data,
                        panel_params,
                        coord,
                        flipped_aes = FALSE) {

    xy <- legend_pos_to_xy(data$legend.position[[1]],
                           panel_params$x.range,
                           panel_params$y.range,
                           inherits(coord, "CoordFlip"))

    richtext_data <- data |>
      dplyr::mutate(
        label = paste0(
          "<span style='color:",
          colour,
          "'>",
          label,
          "</span>"
        ),
        x = xy[1],
        y = xy[2]
      ) |>
      dplyr::group_by(
        PANEL, x, y, size, angle, hjust, vjust,
        alpha, lineheight, family, fontface
      ) |>
      dplyr::summarise(label = paste0(label, collapse = "<br>")) |>
      dplyr::ungroup() |>
      as.data.frame() |>
      dplyr::mutate(colour = "#000000")

    rt_draw_panel(
      richtext_data,
      panel_params,
      coord
    )
  },
  draw_key = ggplot2::draw_key_text,
  required_aes = c(
    "label",
    "colour"
  ),
  default_aes = ggplot2::aes(
    colour = ggtext::GeomRichText$default_aes$colour,
    size = ggtext::GeomRichText$default_aes$size,
    angle = ggtext::GeomRichText$default_aes$angle,
    hjust = 1,
    vjust = 1,
    alpha = ggtext::GeomRichText$default_aes$alpha,
    family = ggtext::GeomRichText$default_aes$family,
    fontface = ggtext::GeomRichText$default_aes$fontface,
    lineheight = 1.2
  )
)

legend_pos_to_xy <- function(legend.position,
                             xrange,
                             yrange,
                             flipped) {
  l <- legend.position

  if(is.numeric(l)) {
    stopifnot("Numeric `legend.position` must have length 2" = length(l) == 2)
    stopifnot("Numeric `legend.position` must have values between 0 & 1" =
                min(l) >= 0 || max(l) <= 1)
    l_num <- l
  }

  if (is.character(l)) {
    stopifnot(l %in% c("left",
                       "right",
                       "bottom",
                       "top",
                       "topright",
                       "bottomright",
                       "bottomleft",
                       "topleft"))
    l_num <- switch(
      l,
      "left" = c(0.025, 0.5),
      "right" = c(0.975, 0.5),
      "bottom" = c(0.5, 0.025),
      "top" = c(0.5, 0.975),
      "topright" = c(0.975, 0.975),
      "bottomright" = c(0.975, 0.025),
      "bottomleft" = c(0.025, 0.025),
      "topleft" = c(0.025, 0.975)
    )
  }
    l_x <- l_num[1] * (xrange[2] - xrange[1]) + xrange[1]
    l_y <- l_num[2] * (yrange[2] - yrange[1]) + yrange[1]

    if (isTRUE(flipped)) {
      return(c(l_y, l_x))
    } else {
      return(c(l_x, l_y))
    }
}

#' Slightly modified from gridtext by Claus O Wilke
#' @keywords internal
#' @noRd
rt_draw_panel <- function(data, panel_params, coord,
                          label.padding = unit(c(
                            0.25,
                            0.25, 0.25, 0.25
                          ), "lines"),
                          label.margin = unit(c(
                            0, 0,
                            0, 0
                          ), "lines"),
                          label.r = unit(0.15, "lines"),
                          na.rm = FALSE) {
  data <- coord$transform(data, panel_params)
  if (is.character(data$vjust)) {
    data$vjust <- compute_just(data$vjust, data$y)
  }
  if (is.character(data$hjust)) {
    data$hjust <- compute_just(data$hjust, data$x)
  }
  gridtext::richtext_grob(data$label, data$x, data$y,
    default.units = "native",
    hjust = data$hjust, vjust = data$vjust, rot = data$angle,
    padding = label.padding, margin = label.margin,
    gp = grid::gpar(
      col = scales::alpha(data$text.colour %||%
        data$colour, data$alpha), fontsize = data$size *
        .pt, fontfamily = data$family, fontface = data$fontface,
      lineheight = data$lineheight
    ),
    r = label.r
  )
}

#' @noRd
`%||%` <- function(a, b) {
  if (!is.null(a)) {
    a
  } else {
    b
  }
}

# From gridtext by Claus O Wilke
#' @noRd
compute_just <- function(just, x) {
  inward <- just == "inward"
  just[inward] <- c("left", "middle", "right")[just_dir(x[inward])]
  outward <- just == "outward"
  just[outward] <- c("right", "middle", "left")[just_dir(x[outward])]

  unname(c(left = 0, center = 0.5, right = 1,
           bottom = 0, middle = 0.5, top = 1)[just])
}

# From gridtext by Claus O Wilke
#' @noRd
just_dir <- function(x, tol = 0.001) {
  out <- rep(2L, length(x))
  out[x < 0.5 - tol] <- 1L
  out[x > 0.5 + tol] <- 3L
  out
}

